ambry-store/src/main/java/com.github.ambry.store/StatsManager.java

```diff
@@ -0,0 +1,186 @@
+        throw new IOException(
+            "Failed to rename " + tempFile.getAbsolutePath() + " to " + statsOutputFile.getAbsolutePath());
+      }
+    } else {
+      throw new IOException("Temporary file creation failed when publishing stats " + tempFile.getAbsolutePath());
+    }
+  }
+
+  /**
+   * Collect and aggregate quota stats from all given {@link Store}s by fetching stats from them sequentially via the
+   * {@link StorageManager}.
+   * @param partitionIds a {@link Set} of {@link PartitionId}s representing a set of {@link Store}s to be fetched
+   * @return a {@link Pair} where the first element is the result in the form of a {@link StatsSnapshot} and the
+   * second element is the number of {@link Store}s that were skipped (either unreachable or an error has occurred).
+   */
+  Pair<StatsSnapshot, List<String>> collectAndAggregate(Set<PartitionId> partitionIds) {
+    StatsSnapshot aggregatedStatsSnapshot = new StatsSnapshot(0L, null);
+    List<String> unreachableStores = new ArrayList<>();
+    for (PartitionId partitionId : partitionIds) {
+      Store store = storageManager.getStore(partitionId);
+      if (store != null) {
+        try {
+          StatsSnapshot statsSnapshot = store.getStoreStats().getStatsSnapshot();
+          aggregate(aggregatedStatsSnapshot, statsSnapshot);
+        } catch (StoreException e) {
+          logger.error("Store exception thrown when getting stats for partitionId: " + partitionId.toString(), e);
+          unreachableStores.add(partitionId.toString());
+        }
+      } else {
+        unreachableStores.add(partitionId.toString());
+      }
+    }
+    return new Pair<>(aggregatedStatsSnapshot, unreachableStores);
+  }
+
+  /**
+   * Performs recursive aggregation of two {@link StatsSnapshot} and stores the result in the first one.
+   * @param baseSnapshot one of the addends and where the result will be
+   * @param newSnapshot the other addend to be added into the first {@link StatsSnapshot}
+   */
+  private void aggregate(StatsSnapshot baseSnapshot, StatsSnapshot newSnapshot) {
+    baseSnapshot.setValue(baseSnapshot.getValue() + newSnapshot.getValue());
+    if (baseSnapshot.getSubtree() == null) {
+      baseSnapshot.setSubtree(newSnapshot.getSubtree());
+    } else if (newSnapshot.getSubtree() != null) {
+      for (Map.Entry<String, StatsSnapshot> entry : newSnapshot.getSubtree().entrySet()) {
+        if (!baseSnapshot.getSubtree().containsKey(entry.getKey())) {
+          baseSnapshot.getSubtree().put(entry.getKey(), new StatsSnapshot(0L, null));
+        }
+        aggregate(baseSnapshot.getSubtree().get(entry.getKey()), entry.getValue());
+      }
+    }
+  }
+
+  /**
+   * Runnable class that collects, aggregate and publish stats via methods in StatsManager.
+   */
+  private class StatsAggregator implements Runnable {
+
+    @Override
+    public void run() {
+      Set<PartitionId> partitionIds = storageManager.getPartitionIds();
```

Comment:
shouldn't you be getting this from the constructor itself? Also, if you do this, your unreachable stores list will be wrong since this set will not contain stores that haven't been started at all because of various errors.